src/mo_values/operator.ml

```diff
@@ -235,8 +235,13 @@ let structural_equality t =
           else
             go (List.find (fun f -> f.T.lab = l1) fs).T.typ v1 v2
     | T.Func (s, c, tbs, ts1, ts2) ->
-        assert (T.is_shared_sort s);
-        fun v1 v2 -> Bool (v1 == v2)  (* HACK *)
+       assert (T.is_shared_sort s);
+       fun v1 v2 -> match v1, v2 with
+       | Tup [Blob _; Text _], Tup [Blob _; Text _] -> Bool (v1 = v2) (* public methods *)
+       | Func _, Tup [Blob _; Text _]
+       | Tup [Blob _; Text _], Func _ -> assert false; (* mixed, cannot determine equality *)
```

Comment:
If equality had access to the actor env then we could probably do better here and maybe even get it right (see above)

Reply:
Yes, but I wouldn't hold my breath.